services/comments.js

```diff
@@ -44,29 +45,87 @@ module.exports = class CommentsService {
 
   /**
    * Adds a tag if it doesn't already exist on the comment.
+   * @throws if tag is already added to the comment
+   * @throws if tag name is not in ALLOWED_TAGS
+   * @param {String} id the id of the comment to tag
+   * @param {String} name the name of the tag to add
+   * @param {String} assigned_by the user id for the user who added the tag
    */
   static addTag(id, name, assigned_by) {
 
-    if (ALLOWED_TAGS.find((t) => t.name === name) == null) {
+    if (ALLOWED_TAGS.find((t) => t === name) == null) {
       return Promise.reject(new Error('tag not allowed'));
     }
 
-    return CommentModel.update({
+    const filter = {
       id,
-      tags: {
-        $ne: {
-          name
+      tags: {$not: {$elemMatch: {name}}}
+    };
+    const update = {
+      $push: {tags: {
+        name,
+        assigned_by,
+        created_at: new Date()
+      }}
+    };
+    return CommentModel.update(filter, update)
+      .then(({nModified}) => {
+        switch (nModified) {
+        case 0:
+
+            // either the tag was already there, or the comment doesn't exist with that id...
+          return this.findById(id)
+              .then(result => {
+                if ( ! result) {
+                  throw new Error(`Can't add tag to comment. There is no comment with id ${id}`);
+                }
+                throw new Error(`Can't add tag ${name} to comment. Comment already has that tag.`);
+              });
+        case 1:
+
+            // tag added
+          return;
+        default:
+          console.warn('CommentsService#addTag modified multiple comments, but it should only have modified 1. '
```

Comment:
This is really an impossible situation due to the unique index on the id field, and the fact that that the update was not made with `{multi: true}`. I'm hesitant to add defensive code for this.